### DISCUSSION TITLE
Automatically test your Orchard Core app's UI

### DISCUSSION THREAD

**Discussion by Piedone** [OrchardCMS/OrchardCore#10376] [Show and tell]
As seen [on TV](https://www.youtube.com/watch?v=mEUg6-pad-E): Everything you need to do UI testing for an Orchard Core app is included in our newly open-sourced [Lombiq UI Testing Toolbox](https://github.com/Lombiq/UI-Testing-Toolbox)! With such UI tests you can automatically verify if everything works in your Orchard app, including checking for server-side or client-side exceptions, HTML markup or accessibility issues. And [thoroughly explained samples](https://github.com/Lombiq/UI-Testing-Toolbox/tree/dev/Lombiq.Tests.UI.Samples) are included. And when you get ahold of it it's kind of... fun?

**Comment by hishamco** [comment]
Seems I need to spend some time to check this on, while I'm interesting on Software Testing in general

**Reply by Piedone** [reply]
Let me know if you have any questions, Hisham!

**Reply by hishamco** [reply]
Is it a VS extension? coz I didn't try it yet

> Let me know if you have any questions, Hisham!

Sure, thanks

**Reply by Piedone** [reply]
It's more of a framework.

**Reply by hishamco** [reply]
I just forked the UI Testing Toolbox & Testing Toolbox, seems the UI included multiple solution. I will have a look and come back to you if there's any questions

Thanks for the HUGE efforts

**Reply by hishamco** [reply]
I just forked the UI Testing Toolbox & Testing Toolbox, seems the UI included multiple solution. I will have a look and come back to you if there's any questions

Thanks for the HUGE efforts

**Reply by Skrypt** [reply]
@piedone What are the advantages over [Cypress](https://www.cypress.io/)? Have you made a comparison between both? And last question, what is it using underneat to do the UI tests? 

There is also Playwright from Microsoft themselves which I would love to find time to take a look at. https://github.com/microsoft/playwright

**Reply by hishamco** [reply]
Seems like Selenium with APIs unifiction for the Web Drivers

**Reply by Piedone** [reply]
Yes, we use Selenium (and Atata), see [the docs here](https://github.com/Lombiq/UI-Testing-Toolbox/blob/dev/Lombiq.Tests.UI/Docs/Tools.md).

While there are several differences between Selenium and Cypress, the one that matters most to me is that you can write Selenium tests in C#. Our UI Testing Toolbox adds a lot of further automated checks and Orchard-specific features.

**Reply by hishamco** [reply]
I'm with you for using Selenium or any framework on top of it like Limbiq Toolbox over Cypress

**Reply by Skrypt** [reply]
One suggestion would be to make it clearer on the repository documentation. Because the first thing you want to know is which framework it uses. We could easily call this module OrchardCore.Selenium.Tests ... and those who wants to use Selenium over Cypress could be interested in your module over our supported one (Cypress). Nothing wrong having multiple implementations of functional tests frameworks ; but for discovery purpose if I search for "Orchard Core + Selenium" in Google, I should get a result pointing to your repository.

**Reply by Piedone** [reply]
Great feedback! I've made it clearer in the docs/description which tools, including Selenium, are used: https://github.com/Lombiq/UI-Testing-Toolbox

**Reply by hishamco** [reply]
I vote to make Selenuim tests as first class citizen along side with Cypress, then we can weight the pros & cons. Again writing unit, functional, integration tests is easy with managed languages

**Reply by Skrypt** [reply]
One thing to not forget. Functional tests can often be required to be written by people who builds frontend stuff. And often, they won't want to build a specific C# module to do so. It's often easier to just use Cypress or Selenium and compose javascript code on the frontend because it doesn't require to be compiled to work. So both frameworks have their advantages/inconvenients.

I think what draws me more to Cypress is it's interactive test runner which works like a charm so far.
